src/hotspot/share/gc/shenandoah/shenandoahFreeSet.cpp

```diff
@@ -42,51 +42,118 @@ ShenandoahFreeSet::ShenandoahFreeSet(ShenandoahHeap* heap, size_t max_regions) :
-//   Remove this function after restructing FreeSet representation.  A problem in the existing implementation is that old-gen
-//   regions are not considered to reside within the is_collector_free range.
-//
-HeapWord* ShenandoahFreeSet::allocate_with_old_affiliation(ShenandoahAllocRequest& req, bool& in_new_region) {
-  ShenandoahAffiliation affiliation = ShenandoahAffiliation::OLD_GENERATION;
+inline bool ShenandoahFreeSet::probe_old_collector_set(size_t idx) const {
+  return _old_collector_free_bitmap.at(idx);
+}
 
-  size_t rightmost = MAX2(_collector_rightmost, _mutator_rightmost);
-  size_t leftmost = MIN2(_collector_leftmost, _mutator_leftmost);
+inline bool ShenandoahFreeSet::in_old_collector_set(size_t idx) const {
+  bool is_old_collector_free = _old_collector_free_bitmap.at(idx);
+  assert (idx < _max, "index is sane: " SIZE_FORMAT " < " SIZE_FORMAT " (left: " SIZE_FORMAT ", right: " SIZE_FORMAT ")",
+          idx, _max, _old_collector_leftmost, _old_collector_rightmost);
+  assert(!is_old_collector_free || has_alloc_capacity(idx), "Old collector free set should contain useful regions");
+  return is_old_collector_free;
+}
 
-  for (size_t c = rightmost + 1; c > leftmost; c--) {
-    // size_t is unsigned, need to dodge underflow when _leftmost = 0
-    size_t idx = c - 1;
-    ShenandoahHeapRegion* r = _heap->get_region(idx);
-    if (r->affiliation() == affiliation && !r->is_humongous()) {
-      if (!r->is_cset() && !has_no_alloc_capacity(r)) {
-        HeapWord* result = try_allocate_in(r, req, in_new_region);
-        if (result != nullptr) {
-          return result;
+inline void ShenandoahFreeSet::add_to_mutator_set(size_t idx) {
+  assert(has_alloc_capacity(idx), "Mutator free set should contain useful regions");
+  assert(!in_old_collector_set(idx) && !in_collector_set(idx), "Freeset membership is mutually exclusive");
+  _mutator_free_bitmap.set_bit(idx);
+}
+
+inline void ShenandoahFreeSet::remove_from_mutator_set(size_t idx) {
+  _mutator_free_bitmap.clear_bit(idx);
+}
+
+inline void ShenandoahFreeSet::add_to_collector_set(size_t idx) {
+  assert(has_alloc_capacity(idx), "Collector free set should contain useful regions");
+  assert(!in_mutator_set(idx) && !in_old_collector_set(idx), "Freeset membership is mutually exclusive");
+  _collector_free_bitmap.set_bit(idx);
+}
+
+inline void ShenandoahFreeSet::remove_from_collector_set(size_t idx) {
+  _collector_free_bitmap.clear_bit(idx);
+}
+
+inline void ShenandoahFreeSet::add_to_old_collector_set(size_t idx) {
+  assert(has_alloc_capacity(idx), "Old collector free set should contain useful regions");
+  assert(!in_mutator_set(idx) && !in_collector_set(idx), "Freeset membership is mutually exclusive");
+  _old_collector_free_bitmap.set_bit(idx);
+}
+
+inline void ShenandoahFreeSet::remove_from_old_collector_set(size_t idx) {
+  _old_collector_free_bitmap.clear_bit(idx);
+}
+
+
+HeapWord* ShenandoahFreeSet::allocate_old_with_affiliation(ShenandoahAffiliation affiliation,
```

Comment:
Will `affiliation` always be `old` here? Should we remove this parameter?

Reply:
Affiliation might be FREE.  First, we try to allocate from within an old-collector-region that is_old().  If that fails, we'll try to allocate from an old_collector_region that is not afilliated.

Reply:
I think the .hpp file declaration documents the intent here:

```
  // Satisfy allocation request req by finding memory that matches affiliation, which either equals req.affiliation
  // or FREE. We know req.is_old().
```

So I think you are right that this interface can probably be simplified and cleaned up (or abstracted and consolidated).

Reply:
I'm adding a comment to explain this.